src/kernel/include/pgtbl.h

```diff
@@ -57,245 +70,398 @@ struct tlb_quiescence {
+	u32_t refcnt_flags;          /* includes refcnt and flags */      /* 1 */
+	struct pgtbl pgtbl;                                               /* 11 */
+	struct cap_pgtbl *parent;    /* if !null, points to parent cap */ /* 1 */
+	u64_t frozen_ts;             /* timestamp when frozen is set. */  /* 2 */
+} __attribute__((packed));
 
-	return ret;
-}
+/* make it an opaque type...not to be touched */
+/* PRY:We change pgtbl_t here. How do we inline these ? */
+/* inline the MPU representation into the pgtbl caps:MPU needs, say, 4 slots, which is 8 words */
+typedef struct pgtbl * pgtbl_t;
 
-/*
- * FIXME: If these need to return a physical address, we should do a
- * va2pa before returning
- */
-static void *
-pgtbl_intern_prune(pgtbl_t pt, u32_t addr)
+///* identical to the capability structure */
+//struct cap_pgtbl {     /* 7 words */
+//	struct cap_header h;                                              /* 1 */
+//	u32_t refcnt_flags;          /* includes refcnt and flags */      /* 1 */
+//	pgtbl_t pgtbl;                                                    /* 1 */
+//	u32_t lvl; 		     /* what level are the pgtbl nodes at? */     /* 1 */
+//	struct cap_pgtbl *parent;    /* if !null, points to parent cap */ /* 1 */
+//	u64_t frozen_ts;             /* timestamp when frozen is set. */  /* 2 */
+//} __attribute__((packed));
+/* PRY:the page table lookup function - look up for a pte, its size order, and flags */
+static u32_t* pgtbl_lkup_pte(pgtbl_t pt, vaddr_t addr, u32_t* size_order, u32_t* flags)
 {
-	unsigned long accum = 0, *pgd;
-	void *page;
-
-	assert(pt);
-	assert((PGTBL_FLAG_MASK & (u32_t)addr) == 0);
-
-	pgd = __pgtbl_lkupan((pgtbl_t)((u32_t)pt|PGTBL_PRESENT), (u32_t)addr >> PGTBL_PAGEIDX_SHIFT, 1, &accum);
-	if (!pgd) return NULL;
-	page = __pgtbl_get((struct ert_intern *)pgd, &accum, 0);
-	accum = 0;
-
-	if (__pgtbl_set((struct ert_intern *)pgd, NULL, &accum, 0)) return NULL;
-
-	return page;
+	u32_t start_addr;
+	u32_t excess_addr;
+	u32_t sz_order;
+
+	/* The pgtbl_t is empty */
+	if(pt==0)
+		return 0;
+
+	/* what is this pt? If this one contains MPU metadata, then we look for its next level */
+	if(COS_PGTBL_TYPE(pt->type_addr)==COS_PGTBL_MPUMETA)
+		pt=COS_PGTBL_METAADDR(pt->type_addr);
+
+	while(1)
+	{
+		/* Get the start address of this */
+		start_addr=COS_PGTBL_STARTADDR(pt->type_addr);
+		/* Start translation */
+		excess_addr=addr-start_addr;
+		sz_order=COS_PGTBL_SIZEORDER(pt->type_addr);
+		/* Not in range */
+		if((excess_addr>>sz_order)>8)
+			return 0;
+
+		/* See if this entry exists */
+		if((pt->data[excess_addr>>sz_order]&PGTBL_PRESENT)==0)
+			return 0;
+
+		if((pt->data[excess_addr>>sz_order]&PGTBL_COSFRAME)==0)
+			return 0;
+
+		/* See what is in this entry. Is this a pte or an internal node */
+		if((pt->data[excess_addr>>sz_order]&PGTBL_LEAF)!=0)
+			break;
+
+		/* Must be an internal node */
+		pt=COS_PGTBL_STARTADDR(pt->type_addr);
+	}
+	*size_order=sz_order;
```

Comment:
Again, representation should be hidden from the calling sites.  Just return size and order as separate variables if they are logically separate.